### ISSUE TITLE
Duplicate case in switch (compile error) when building without cgo

### ISSUE THREAD

**Issue by bartolomej** [onflow/flow-go#5899]
## Problem Definition

I'm trying to build flow-emulator for JS/WASM target, but I'm running into this error when [building without cgo](https://github.com/onflow/crypto?tab=readme-ov-file#build-without-cgo):

```bash
# wasm/main.go is just my entry point that imports flow-go and flow-emulator
$ CGO_ENABLED=0 GOOS=js GOARCH=wasm go build -tags=no_cgo -o flow-emulator.wasm wasm/main.go
# github.com/onflow/flow-go/module/signature
../../go/pkg/mod/github.com/!the-!k-!r-!o-!k/flow-go@v0.34.0-crescendo-preview.9.0.20240508085818-dd583cbc9a7b/module/signature/type_encoder.go:68:7: duplicate case 2 * SigLen (constant 0 of type int) in expression switch
        ../../go/pkg/mod/github.com/!the-!k-!r-!o-!k/flow-go@v0.34.0-crescendo-preview.9.0.20240508085818-dd583cbc9a7b/module/signature/type_encoder.go:66:7: previous case
```

Looks like `SigLen` is 0 when building without cgo - I believe it's because C source isn't included/compiled in the build, so I'm guessing it just defaults to 0?

https://github.com/onflow/flow-go/blob/e4bdb7ea0603cbb11a4e5605eb6695be5c7c7c98/module/signature/type_encoder.go#L65-L70

## Proposed Solution

Is there a simple fix we can do to take into account this edge case? Maybe using if/else-if statement instead of switch doesn't cause the same compiler error or something like that?

## Definition of Done

flow-go compiles successfully with command such as:

```
CGO_ENABLED=0 go build -tags=no_cgo cmd/execution/main.go
```

**Comment by turbolent** [comment]
Seems odd that `SigLen` ends up being 0 🤔

cc @tarakby

**Comment by tarakby** [comment]
Maybe @jribbink knows more about the flow-emulator build for wasm?

**Comment by bartolomej** [comment]
@tarakby I guess I just assumed that flow-go is also built in a way that it doesn't require cgo. Didn't realise it says that in the README.

> I believe the flow-go repo always requires BLS in some way...

After quickly looking into the codebase, it seems like this is true. One example is the `PublicKey.verifyPoP` function exposed to Cadence. 

So it looks like we can't build a WASM emulator that fully matches the non-WASM emulator in feature completeness without cgo enabled. But it doesn't work to enable CGO when building for wasm/js target, since that's just not yet supported by Go natively.

Related discussion: https://stackoverflow.com/questions/54946214/cgo-library-build-to-js-wasm-file

The above discussion does however mention there seems to be a way to use cgo and also compile C code to wasm (or at least that Qt somehow solved that limitation), but that sounds pretty hacky and complicated, so not sure if it's worth digging into that.

So in summary, I believe we could:
- hope that WASM flow-emulator still works good enough that it would be useful, even without BLS (and warn users of some features not working)
- see if there is some way to import C code to Go as a wasm binary instead of a native OS executable (may be a difficult task)

For context, I'm trying to build a PoC of the new fully browser-based playground. We also made this grant proposal: https://github.com/onflow/developer-grants/issues/260

**Comment by tarakby** [comment]
Thanks for the context! 

- flow-go` is the Flow distributed protocol repo, and the protocol heavily relies on BLS, so I don't think it will be possible to make it build without BLS (and therefore without cgo). I know your message doesn't challenge this, but I'm recalling it before the next points :)
- the `flow-emulator` repo however, does not implement the distributed protocol (or implements a dummy simulated version of the distributed protocol), so it can make sense to build it without BLS (and cgo). Now that the crypto library has a `no-cgo` mode, this should be even easier. Any call to a BLS-related function or constant should be replaceable in flow-emulator, using the Go build tags([example](https://github.com/onflow/crypto/blob/main/no_cgo.go#L1), we can also use the native tag `wasm`).
- From your summary points, the first point seems more reasonable. A wasm-flow-emulator would exclude the Cadence BLS functions. There are still other cryptographic algorithms to use on Cadence so I don't think any developer would miss BLS IMO.
- I think we can close this issue and open an issue on the flow-emulator repo to make it wasm buildable (@turbolent or @jribbink feel free to suggest something else)

**Comment by bartolomej** [comment]
@tarakby Upon discussing this with @bluesign, looking at the database, and reading your response I think I agree.

So in summary:
- Since the flow-emulator doesn't depend on any protocol code in flow-go, BLS won't need to be used for things like consensus
- I think the only place where BLS is used within the emulator is in the Cadence crypto library (exposed in [fvm/environment/crypto_library.go](https://github.com/onflow/flow-go/blob/master/fvm/environment/crypto_library.go)).
  - But this also shouldn't be a big issue, since we can just disallow usage of BLS crypto algorithms (for hashing and signing) when doing things like creating account keys.
  - In case someone does use BLS functions in their Cadence code, we could also detect that on our end and report an error. Probably wouldn't be a deal breaker for most people as you mentioned.

**Comment by bartolomej** [comment]
@tarakby @turbolent However for building flow-emulator without cgo, this "duplicate case in switch" compiler error still prevent us from having a successful build. But since that is our concern, I'm thinking of just using my forked repo (where I've fixed the issue already) instead of this one, which would come at a slight cost of having to keep that up-to-date. Do you agree this is the way to go?

**Comment by tarakby** [comment]
> just disallow usage of BLS crypto algorithms (for hashing and signing) when doing things like creating account keys (in flow-emulator / flow-cli / UI).

The good news is that BLS can't be used as an account key algorithm, only ECDSA can. BLS in isn't used by CLI, and can only be used by the flow-emulator if used in a Cadence contract (a niche use-case as you mentioned). 

I tried out the fix and noticed it's enough to make flow-emulator build. I initially thought the emulator requires multiple fixes and the one you pointed is only the first error. I opened this [PR](https://github.com/onflow/flow-go/pull/5906) in flow-go, feel free to open a PR on flow-emulator to import the new flow-go version.
That would avoid you relying on a fork. In general, feel free to open PRs against the main repos, they will get accepted as long as they don't change the main case behaviour.
